DOCUMENTATION.md

```diff
@@ -43,6 +41,314 @@ The overall territory is laid out here in this [C4](https://c4model.com) System
+   - Implement a leasing system akin to that implemented in the CosmosDb Change Feed Processor library to manage balanced assignment of Tranches to Processor Instances. As leases are assigned and revoked, the Source needs to run a Reader and Ingester per assigmnent, that reads the underlying data and forwards the relevant data solely for the batches assigned to the specific Processor Instance in question
+
+<a name="why-not-queue-all-the-things"></a>
+## Processing Reactions based on a Store's Notifications vs Just using a Queue
+
+Its not uncommon for people to use a Queue (e.g. RabbitMQ, Azure Service Bus Queues, AWS SQS etc) to manage reactive processing. This typically involves wiring things up such that relevant events from the Store's notifications trigger a message on that queue. 
+
+As a system grows in features, Reaction processing tends to grow into a workflow, with complex needs regarding being able to deal with partner system outages, view the state of a workflow, cancel complex processes etc. That's normally not now it starts off though, especially in any whitepaper trying to sell you a Queue, Low Code or FaaS. 
+
+Just using a Queue has a number of pleasing properties:
+- The only thing that can prevent your Store notifications processing from getting stuck is if your Queue becomes unwriteable. Typically it'll take a lot of activity in your system for writing to the queue to become the bottleneck.
+- You have a natural way to cope with small spikes in traffic
+- You can normally easily rig things such that you run multiple instances of your Queue Processor if the backlog gets too long (i.e. the Competing Consumer pattern)
+- Anything that fails can be deferred for later retry pretty naturally (i.e. set a visibility timeout on the item)
+- Any case which for some reason triggers an exception, can be dealt with as a Poison Message
+- After a configured amount of time or attempts, you can have messages shift to a Dead Letter Queue. There are established practices regarding alerting in such circumstances and/or triggering retries of Dead Lettered messages  
+- Arranging processing via a Queue also gives a natural way for other systems (or ad-hoc admin tools) to trigger such messages too. For example, you can easily do a manual trigger or re-trigger of some reaction by adding a message out of band.
+- Queues are a generally applicable technique in a wide variety of systems, and look great on a system architecture diagram or a resume
+
+There are also a number of potential negatives; the significance of that depends on your needs:
+- You have another system to understand, worry about the uptime of, and learn to monitor
+- Marking a message invisible (or running competing consumers) will mean you can receive messages regarding a particular action out of order. In other words, you get no useful ordering guarantees (frequently this is not a problem)
```

Comment:
> receive messages regarding a particular action out of order

SQS FIFO would like to have a word with you. You could ship messages with `GroupId=StreamName` and get the same ordering guarantees as propulsion 😉

Reply:
Caveats about limits of course (there's a limit of in-flight messages)

Reply:
yes Lambda driven from DDB Streams can do cool things that way too
Azure Service Bus Queues have that too
But in most cases the concurrency limits are not interesting, are they?

If you can formulate a sentence to ack this as a possibiltiy with limitations, that'd be excellent. (or do you have an example?)

Reply:
> But in most cases the concurrency limits are not interesting, are they?

I've never had them be an issue.

> If you can formulate a sentence to ack this as a possibiltiy with limitations, that'd be excellent. (or do you have an example?)
Something like:
```suggestion
- Your choice of queue will affect the ordering guarantees you get. Some queues like SQS offer no useful ordering guarantees whatsoever, this means your processor can receive events about a particular action out of order. Other queues like Azure Service Bus, and SQS FIFO allow FIFO processing based on group-keys but do so at a significantly lower throughput than their orderless counterparts.
```

Reply:
looking good, thanks!

Reply:
Applied with slight rewordings in https://github.com/jet/propulsion/pull/200/commits/04b8b7a56fc6e89f4885f6b9851a11d9b40c54cc